src/process_util.rs

```diff
@@ -97,6 +101,66 @@ pub fn run_and_wait(cmd: &mut Command)
     unreachable!();
 }
 
+pub fn run_with_tty_and_wait(cmd: &mut Command, tty_fd: i32)
+    -> Result<i32, String>
+{
+    let pgid = unsafe { getpgrp() };
+    let child = match cmd.spawn() {
+        Ok(child) => child,
+        Err(e) => return Err(format!("Error running {:?}: {}", cmd, e)),
+    };
+    let mut trap = Trap::trap(&[SIGCHLD, SIGTERM]);
+    'signal_loop: for signal in trap.by_ref() {
+        match signal {
+            SIGCHLD => {
+                debug!("Received SIGCHLD signal, waiting child");
+                loop {
+                    match waitpid(child.pid(), Some(WNOHANG | WUNTRACED)) {
+                        Ok(WaitStatus::Stopped(_, signum)) => {
+                            if signum == SIGTTOU || signum == SIGTTIN {
+                                try!(give_tty_to(tty_fd, child.pid()));
+                                try!(child.signal(SIGCONT)
+                                     .map_err(|e| format!(
+                                         "Error when sending SIGCONT
+                                          signal to {:?}: {}", cmd, e)));
+                            }
+                            continue 'signal_loop;
```

Comment:
SIGCHLD is old-style signal, which means if it raised multiple times it is signalled only once. Then it's expected that you will waitpid() all the children before waiting for the next signal.